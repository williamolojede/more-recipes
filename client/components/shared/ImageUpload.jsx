import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactLoading from 'react-loading';

class ImageUpload extends Component {
  render() {
    const {
      imgUrl,
      altText,
      imageUploadProgress,
      className,
      defaultImgUrl,
      imageUploadHandler
    } = this.props;

    return (
      <div className={`image-upload ${className}`}>
        <div className="image-upload__preview">
          <img
            src={imgUrl || defaultImgUrl}
            alt={altText}
            className="image-upload__preview___image"
          />
          <div className="image-upload__progress">
            {
              imageUploadProgress === 'started' &&
              <ReactLoading
                type="spokes"
                color="#000"
                height="65px"
                width="65px"
                className="loading-icon"
              />
            }
          </div>
        </div>
        <label htmlFor="recipe-image" className="z-depth-1">Upload new picture</label>
        <input
          type="file"
          id="recipe-image"
          name="filename"
          accept="image/gif, image/jpeg, image/png"
          ref={node => this.recipeImageInput = node}
          onChange={() => imageUploadHandler(this.recipeImageInput)}
        />
      </div>
    );
  }
}

ImageUpload.defaultProps = {
  className: '',
  imgUrl: ''
};

ImageUpload.propTypes = {
  className: PropTypes.string,
  imgUrl: PropTypes.string,
  altText: PropTypes.string.isRequired,
  defaultImgUrl: PropTypes.string.isRequired,
  imageUploadProgress: PropTypes.oneOf(['pending', 'started', 'failed', 'completed']).isRequired,
  imageUploadHandler: PropTypes.func.isRequired
};

export default ImageUpload;
