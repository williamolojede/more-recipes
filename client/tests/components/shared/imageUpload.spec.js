import React from 'react';
import ImageUpload from '../../../components/shared/ImageUpload';
import defaultConfig from '../../../config/default';

function setup() {
  const props = {
    altText: 'beautiful image',
    defaultImgUrl: defaultConfig.userImgUrl,
    imageUploadProgress: 'pending',
    imageUploadHandler: jest.fn()
  };
  const wrapper = mount(<ImageUpload {...props} />);

  return {
    wrapper
  };
}

describe('Image Upload', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a loading component when imgageUploadProgress is started', () => {
    const { wrapper } = setup();
    wrapper.setProps({ imageUploadProgress: 'started' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call a function to handle image upload when file is selected with input field', () => {
    const { wrapper } = setup();
    const uploadInput = wrapper.find('#recipe-image');
    const imageUploadHandlerSpy = jest.spyOn(wrapper.instance().props, 'imageUploadHandler');
    expect(uploadInput.length).toBe(1);
    uploadInput.simulate('change');
    expect(imageUploadHandlerSpy).toHaveBeenCalledTimes(1);
  });
});
