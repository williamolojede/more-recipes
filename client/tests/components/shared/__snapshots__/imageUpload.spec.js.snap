// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Upload should render a loading component when imgageUploadProgress is started 1`] = `
<ImageUpload
  altText="beautiful image"
  className=""
  defaultImgUrl="https://catholic-foundation.org/wp-content/uploads/2017/07/unisex-avatar.png"
  imageUploadHandler={[Function]}
  imageUploadProgress="started"
  imgUrl=""
>
  <div
    className="image-upload "
  >
    <div
      className="image-upload__preview"
    >
      <img
        alt="beautiful image"
        className="image-upload__preview___image"
        src="https://catholic-foundation.org/wp-content/uploads/2017/07/unisex-avatar.png"
      />
      <div
        className="image-upload__progress"
      >
        <Loading
          className="loading-icon"
          color="#000"
          delay={1000}
          height="65px"
          type="spokes"
          width="65px"
        >
          <div
            className="loading-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg class=\\"icon-blank\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"></svg>
",
              }
            }
            style={
              Object {
                "fill": "#000",
                "height": "65px",
                "width": "65px",
              }
            }
          />
        </Loading>
      </div>
    </div>
    <label
      className="z-depth-1"
      htmlFor="recipe-image"
    >
      Upload new picture
    </label>
    <input
      accept="image/gif, image/jpeg, image/png"
      id="recipe-image"
      name="filename"
      onChange={[Function]}
      type="file"
    />
  </div>
</ImageUpload>
`;

exports[`Image Upload should render with right amount of elements 1`] = `
<ImageUpload
  altText="beautiful image"
  className=""
  defaultImgUrl="https://catholic-foundation.org/wp-content/uploads/2017/07/unisex-avatar.png"
  imageUploadHandler={[Function]}
  imageUploadProgress="pending"
  imgUrl=""
>
  <div
    className="image-upload "
  >
    <div
      className="image-upload__preview"
    >
      <img
        alt="beautiful image"
        className="image-upload__preview___image"
        src="https://catholic-foundation.org/wp-content/uploads/2017/07/unisex-avatar.png"
      />
      <div
        className="image-upload__progress"
      />
    </div>
    <label
      className="z-depth-1"
      htmlFor="recipe-image"
    >
      Upload new picture
    </label>
    <input
      accept="image/gif, image/jpeg, image/png"
      id="recipe-image"
      name="filename"
      onChange={[Function]}
      type="file"
    />
  </div>
</ImageUpload>
`;
